### PR TITLE
spirv-diff: Basic support for OpTypeForwardPointer 

### PR DESCRIPTION
**Pull request by ShabbyX**
Currently, the diff tool matches types bottom up, so on every
instruction it expects to know if its operands are already matched or
not.  With cyclical references, it cannot know that.  Type matching
would need significant rework to be able to support such a use case; for
example, it may need to maintain a set of plausable matches between type
pointers that are forward-referenced, and potentially back track when
later the types turn out to be incompatible.

In this change, OpTypeForwardPointer is supported in the more common and
trivial case.  Firstly, forwarded type pointers are only matched if they
have they have the same storage class and point to the same type opcode:

- In the presence of debug info, matching is done only if the names are
  unique in both src and dst.
- In the absence of debug info, matching is done only if there is only
  one possible matching.

Fixes: #4754

### CODE REVIEW

**Comment by ShabbyX** [comment]
@s-perron Refactoring change has a separate PR that should land first: https://github.com/KhronosGroup/SPIRV-Tools/pull/4760

File: source/diff/diff.cpp [hunk 1]
```diff
@@ -793,6 +814,25 @@ uint32_t Differ::GroupIdsHelperGetTypeId(const IdInstructions& id_to,
...
+  assert(inst && inst->opcode() == SpvOpTypePointer);
+  return SpvStorageClass(inst->GetInOperand(0).words[0]);
+}
+
+SpvOp Differ::GroupIdsHelperGetTypePointerTypeOp(const IdInstructions& id_to,
+                                                 uint32_t id) {
+  const opt::Instruction* inst = GetInst(id_to, id);
+  assert(inst && inst->opcode() == SpvOpTypePointer);
+
+  const uint32_t type_id = inst->GetInOperand(1).AsId();
```
Comment (s-perron) [inline source/diff/diff.cpp:829]: Same here.

File: source/diff/diff.cpp [hunk 2]
```diff
@@ -556,6 +569,14 @@ void IdInstructions::MapIdsToInfos(
       case SpvOpMemberDecorate:
         info_map = &decoration_map_;
         break;
+      case SpvOpTypeForwardPointer: {
+        uint32_t id = inst.GetOperand(0).AsId();
```
[the remark this answers is not in the document]
Comment (ShabbyX) [reply source/diff/diff.cpp:573]: Oh nice, I missed `GetSingleWord*Operand`. I did a pass over the whole file and switched to that in a separate PR to keep this diff smaller: https://github.com/KhronosGroup/SPIRV-Tools/pull/4768

[the remark this answers is not in the document]
Comment (ShabbyX) [reply source/diff/diff.cpp:573]: I did change these cases too per your recommendation (not just the ones that do `.words[0]`), though I'm not sure that is strictly better. With `AsId()` there's an assert() that the operand is indeed an id, which is missed by this change.

File: source/diff/diff.cpp (hunk 1 above)
Comment (ShabbyX) [reply source/diff/diff.cpp:829]: Done.

File: source/diff/diff.cpp [hunk 3]
```diff
@@ -1967,6 +2027,108 @@ void Differ::MatchExecutionModes() {
...
+  // info, with minimal matching of the type being forwarded:
+  //
+  // - Group by class
+  //   - Group by OpType being pointed to
+  //     - Group by debug info
+  //       - If same name and unique, match
+  //     - If leftover is unique, match
+
+  // Group forwarded pointers by storage class first.
+  GroupIdsAndMatch<SpvStorageClass>(
```
[the remark this answers is not in the document]
Comment (ShabbyX) [reply source/diff/diff.cpp:2112]: I did some refactoring, let me know how it looks to you know. I see `GroupIdsAndMatch` as a `for (...)` loop if that helps readability (updated comment to make it clear it's a loop). I kept the two outer `GroupIdsAndMatch`es to keep it clear how the instructions are divided up for matching (storage class and type op). The details afterwards (match by name, match unique ones) are refactored to helper functions.

**Review by s-perron — APPROVED** [review]

**Comment by s-perron** [comment]
I've rebased the change.  If all of the tests pass, I'll merge the change.

**Comment by ShabbyX** [comment]
Thanks!

**Pull request merged**